### ISSUE TITLE
Add view(array, inds) with inds::NTuple

### ISSUE THREAD

**Issue by juliohm** [JuliaLang/julia#46576]
It would be nice if we could do `view(array, inds)` with `inds::NTuple`. Right now we have code that is dependent on `SVector` from StaticArrays.jl just because the language doesn't support this view with static indices: https://github.com/JuliaGeometry/Meshes.jl/blob/d927d5629f5f676684cb2b05ee35c4de9379938c/src/connectivities.jl#L99

**Comment by mcabbott** [comment]
What do you picture this returning? E.g. for `view(A, (1,2,3))`, and `view(A, (1:2, 2:3))`, where say `A = rand(10)`.

**Comment by mcabbott** [comment]
But what exactly does it return? If a tuple, then this is `getindex.((A,), (1,2,3))`, so why call it `view`?

**Comment by juliohm** [comment]
I don't know if I understand the example. I'm just mentioning the use case
where we need to create a view of a large buffer Vector at a few specific
integer locations. Concretely:

view([1,2,3,...,100], (2,5,9)) should return the view([1,2,3,...,100],
[2,5,9]) without allocating these tiny vectors in hot loops.

In this use case I linked in Meshes.jl we materialize thousands of
geometries in a mesh by just creating views into a buffer of coordinates.

Em qui., 1 de set. de 2022 13:11, Michael Abbott ***@***.***>
escreveu:

**Comment by mikmoore** [comment]
I think this is unnecessary? It seems that
```julia
x = (1:5).^2
y = view.(Ref(x),(2,5))
eltype(y) # SubArray{Int64, 0, Vector{Int64}, Tuple{Int64}, true}
```
does most of what you would want, returning a tuple of zero-dimensional views into `x`. I suppose the piece missing here is that you would have to access `y` as `y[i][]` to get the value rather than the view. But as you pointed out, your use case is covered by `StaticArrays` which seems to be a totally sensible dependency for this sort of thing.

**Comment by juliohm** [comment]
I think it is still necessary, it is very convenient at least. Being able to take views with tuples of integers seems like a common use case in these types of optimized code, and depending on a complex struct such as SVector just for that is suboptimal. SVector comes with a lot of behavior we don't need.

**Comment by thchr** [comment]
That particular method from Meshes should probably just be constructing a tuple directly rather than bothering with views though? E.g., it could be something like:
```
function materialize(connec::Connectivity{PL,N},
                              points::AbstractVector{P}) where {PL<:Polytope,N,P<:Point}
    #PL(view(points, SVector(connec.indices...)))
    PL(ntuple(i->points[connect.indices[i]], Val(N)))
end
```
Which would not (heap) allocate anything.

**Comment by juliohm** [comment]
The problem with tuple is that it allocates, or am I'm missing something?
We have a large vector of points and need to just view it in specific
entries to build triangles and other geometric objects. I thought that
viewing into the large array was the non-allocating approach. Instantiating
a tuple will still allocate right?

Em sáb., 3 de set. de 2022 00:32, Thomas Christensen <
***@***.***> escreveu:

**Comment by thchr** [comment]
Tuples don't generally allocate (i.e. they are on the stack/registers; equivalent to just defining `N` local variables).

**Comment by juliohm** [comment]
The tuples are not the problem, the problem is the view of the large vector
of points. At the end we will be instantiating new tuples of points not
existent anywhere else. Makes sense?

Em sáb., 3 de set. de 2022 09:07, Thomas Christensen <
***@***.***> escreveu:

**Comment by mikmoore** [comment]
A tuple of numbers requires the same memory (for the largest native types -- less for smaller) as a tuple of indices that describe a view of the same size, but doesn't need a pointer to the parent array like the view does. Recall that the view needs to remember what indices you wanted to reference. The advantage is that the tuple of values is already loaded and ready to go (and on the stack or in registers, aligned for SIMD, etc) while the view must frequently dereference that pointer to access values.

In short, a tuple of values is superior to a view with a tuple of indices. The only benefit the view offers is the ability to mutate the parent's contents.

**Comment by juliohm** [comment]
> In short, a tuple of values is superior to a view with a tuple of indices. The only benefit the view offers is the ability to mutate the parent's contents.

@mikmoore I don't know if I understand what you mean.

Here is a concrete example:

```julia
points = rand(Point3, 100)

inds1 = (1,4,10,5) # tuple
inds2 = [1,4,10,5] # vector

@allocated view(points, collect(inds1)) # allocates 144
@allocated view(points, inds2) # allocates 48
```

Now imagine that I am taking these views thousands of times. If the view worked with tuple directly I could pass `inds1` as is to the view and avoid the allocations, which is what is happening with `inds2`. Now I need to convert Tuple to Vector manually using `collect` just to take a view.

What am I missing?

**Comment by thchr** [comment]
I think the idea here is that you could just store a tuple `(points[i1], points[i2], ...)` rather than storing indices `inds = (i1, i2, ...)`. That way, you don't need to bother with views at all.

**Comment by mikmoore** [comment]
You're getting pushback partly because you haven't provided a mutlidimensional generalization of this concept, which `view` otherwise supports. It would also require a significant revision to `SubArray` to accomplish something that is easy to accomplish other ways -- one of which is through the existing, robust, and widely-popular `StaticArrays.jl`.

The earlier suggestion of `x = view.(Ref(points), inds1)` (returning a tuple of 0-dimensional views) appears to do exactly what you want if you can suffer through an extra level of indirection `x[i][]` to access your points. If that's too inconvenient, then it's very easy to define your own type that handles that for you using the [array interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array). To show how easy this is, I've done it here:
```julia
struct TupleView{T,N,S} <: AbstractVector{T}
	v::NTuple{N,S}
end

function TupleView(arr::AbstractArray{T},tup::NTuple{N}) where {T,N}
	v = view.(Ref(arr),tup)
	S = eltype(v)
	return TupleView{T,N,S}(v)
end

Base.size(t::TupleView) = (length(t.v),)
Base.IndexStyle(::Type{<:TupleView}) = IndexLinear()
Base.getindex(t::TupleView, i::Int) = getindex(t.v[i])
Base.setindex!(t::TupleView, x, i::Int) = setindex!(t.v[i], x)
```
You'll have to forgive me for not installing `Meshes.jl` for the demonstration. Instead I'm just assuming that a `NTuple{2,Float64}` is a sufficiently complicated eltype that your `Point3` will also work.
```julia
julia> using BenchmarkTools

julia> points = tuple.(rand(5),rand(5))
5-element Vector{Tuple{Float64, Float64}}:
 (0.24355809849339638, 0.1545656218185506)
 (0.999385776206506, 0.74694329233283)
 (0.7084560646541075, 0.2499990972839683)
 (0.6083558078986703, 0.6652643086199085)
 (0.10627917652354268, 0.12032731844639666)

julia> inds = (1,3,4)
(1, 3, 4)

julia> tv = @btime TupleView($points,$inds)
  4.700 ns (0 allocations: 0 bytes)
3-element TupleView{Tuple{Float64, Float64}, 3, SubArray{Tuple{Float64, Float64}, 0, Vector{Tuple{Float64, Float64}}, Tuple{Int64}, true}}:
 (0.24355809849339638, 0.1545656218185506)
 (0.7084560646541075, 0.2499990972839683)
 (0.6083558078986703, 0.6652643086199085)

julia> tv[2] = (0.0,0.0)
(0.0, 0.0)

julia> points
5-element Vector{Tuple{Float64, Float64}}:
 (0.24355809849339638, 0.1545656218185506)
 (0.999385776206506, 0.74694329233283)
 (0.0, 0.0)
 (0.6083558078986703, 0.6652643086199085)
 (0.10627917652354268, 0.12032731844639666)
```

Some re-architectonicing of `Meshes.jl`, or simply copying points instead of indexing might remove the need for this entirely. Note that while there may be a memory space cost to copying the points, there is probably a speed benefit due to gathering them into a small block of memory. I wouldn't discount the copy entirely.

**Comment by thchr** [comment]
> @thchr but that doesn't work in the context of Meshes.jl. We have a global vector of vertices and only store the indices of connectivities to save memory.

You can roll this conversion directly then (or use @mikmoore's suggestion if you need the result to behave like a vector):
```jl
julia> tupleview(v, inds::NTuple{N,Int}) where N = ntuple(i->v[inds[i]], Val(N))

julia> v = rand(100)
julia> inds = (1,3,4)
julia> @btime tupleview($v, $inds)
  1.700 ns (0 allocations: 0 bytes)
(0.02597032912898689, 0.6171268026598445, 0.8991203209161366)
```

With StaticArrays, this could be:
```
julia> saview(v, inds::NTuple{N,Int}) where N = SVector{N,eltype(v)}(v[ind] for ind in inds)
```
Which compiles to the same thing as `tupleview`.

**Comment by juliohm** [comment]
Reading this thread again I now fully understand the counter-proposal. Thanks for sharing it, I will try to refactor the code to construct tuples of points instead of views.
